src/types/basic/uint.ts

```diff
@@ -123,6 +129,101 @@ export class NumberUintType extends UintType<number> {
   }
 }
 
+const TWO_POWER_32 = 2 ** 32;
+
+/**
+ * For 64 bit number, we want to operator on HashObject
+ * over bytes to improve performance.
+ */
+export class Number64UintType extends NumberUintType {
+  constructor() {
+    super({byteLength: 8});
+    this._typeSymbols.add(NUMBER_64_UINT_TYPE);
+  }
+
+  // TODO: move this logic all the way to persistent-merkle-tree
```

Comment:
right now for `tree_applyUint64Delta` scenario we have to:
+ traverse the tree based on gindex
+ deserialize, apply delta
+ serialize, traverse the tree to apply the result

technically we can move this logic all the way to persistent-merkle-tree to traverse the tree only 1 time, but is it weird to do that?

Reply:
Oh right, we 100% should only read each chunk once. Can you add a utility function to set deltas in batch such that each tree leaf chunk is read only once?

Reply:
> Even at the epoch transition we will likely recreate the entire tree. So can we have a fast utility to do that with minimal tree traversing and touching each chunk exactly once?

for epoch transition, I already traverse each chunk once by leveraging `getNodesAtDepth()` api, please have a look at `number64_newTreeFromUint64Deltas()` implementation.

My concern was mainly for the `tree_applyUint64Delta` flow, in lodestar it'd improve the `processBlock()` flow where we have to apply deltas for sync committee balances.

Initially I thought it's weird having to bring this logic to persistent-merkle-tree, but we could extend the Tree class and implement it right in ssz, and use the new Tree here.

> Oh right, we 100% should only read each chunk once. Can you add a utility function to set deltas in batch such that each tree leaf chunk is read only once?

👍  I'd love to implement that in a separate PR